url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+            getCPSR().setValue(getSPSR());
+        } else if (s) {
+            getCPSR().setNBit(BitOp.getBit32(dest, 31));
+            getCPSR().setZBit(dest == 0);
+            getCPSR().setCBit(IntegerExt.carryFrom(left, right));
+            getCPSR().setVBit(IntegerExt.overflowFrom(left, right, true));
+        }
+
+        setReg(rd, dest);
+    }
+
+    /**
+     * キャリー付き加算命令。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeALUAdc(InstructionARM inst, boolean exec) {
+        int id = inst.getOpcodeSBitShiftID();
+        boolean s = inst.getSBit();
+        int rn = inst.getRnField();
+        int rd = inst.getRdField();
+        int opr = getAddrMode1(inst);
+        int left, center, right, dest;
+
+        if (!exec) {
+            printALUDisasm(inst, id);
+            return;
+        }
+
+        if (!inst.satisfiesCond(getCPSR())) {
+            return;
+        }
+
+        left = getReg(rn);
+        center = opr;
+        right = BitOp.toInt(getCPSR().getCBit());
+        dest = left + center + right;
+
+        if (s && rd == 15) {
+            getCPSR().setValue(getSPSR());
+        } else if (s) {
+            int lc = left + center;
+            boolean lc_c = IntegerExt.carryFrom(left, center);
+            boolean lc_v = IntegerExt.overflowFrom(left, center, true);
+
+            getCPSR().setNBit(BitOp.getBit32(dest, 31));
+            getCPSR().setZBit(dest == 0);
+            getCPSR().setCBit(lc_c || IntegerExt.carryFrom(lc, right));
+            getCPSR().setVBit(lc_v || IntegerExt.overflowFrom(lc, right, true));
+        }
+
+        setReg(rd, dest);
+    }
+
+    /**
+     * キャリー付き減算命令。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeALUSbc(InstructionARM inst, boolean exec) {
+        int id = inst.getOpcodeSBitShiftID();
+        boolean s = inst.getSBit();
+        int rn = inst.getRnField();
+        int rd = inst.getRdField();
+        int opr = getAddrMode1(inst);
+        int left, center, right, dest;
+
+        if (!exec) {
+            printALUDisasm(inst, id);
+            return;
+        }
+
+        if (!inst.satisfiesCond(getCPSR())) {
+            return;
+        }
+
+        left = getReg(rn);
+        center = opr;
+        right = BitOp.toInt(!getCPSR().getCBit());
+        dest = left - center - right;
+
+        if (s && rd == 15) {
+            getCPSR().setValue(getSPSR());
+        } else if (s) {
+            int lc = left - center;
+            boolean lc_c = IntegerExt.borrowFrom(left, center);
+            boolean lc_v = IntegerExt.overflowFrom(left, center, false);
+
+            getCPSR().setNBit(BitOp.getBit32(dest, 31));
+            getCPSR().setZBit(dest == 0);
+            getCPSR().setCBit(!(lc_c || IntegerExt.borrowFrom(lc, right)));
+            getCPSR().setVBit(lc_v || IntegerExt.overflowFrom(lc, right, false));
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1487:1498, 1531:1542.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.